__tests__/undo-manager.test.ts

```diff
@@ -568,3 +568,31 @@ describe("includeHooks flag", () => {
         ])
     })
 })
+
+test("#15 - rollback by recorder.undo() should not be an UndoState", () => {
```

Comment:
**issue:** @songzhenqi - this test passes even if I revert your changes in `src/undo-manager.ts`.

**suggestion:** can you please write a test that would fail without your changes, so that we can demonstrate the intention behind it, and also protect against regressions?

Reply:
@songzhenqi - is it possible you have some local changes that haven't made it up to this branch? When I check out `songzhenqi:fix-rollback-from-action-error`, and then:

1. `yarn build`
1. `yarn test`

I get the following output, even if I have commented out `call.env = undefined` in `src/undo-manager.ts`:

<details>

```
$ yarn test:dev && yarn test:prod
$ cross-env NODE_ENV=development JEST_JUNIT_OUTPUT=../../test-results/mst-middlewares/dev.xml yarn jest
$ jest
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
 PASS   test  __tests__/action-logger.test.ts
 PASS   test  __tests__/simple-action-logger.test.ts
 PASS   test  __tests__/time-traveller.test.ts
 PASS   test  __tests__/custom.test.ts
 PASS   test  __tests__/undo-manager.test.ts
 PASS   test  __tests__/atomic.test.ts
 PASS   test  __tests__/redux.test.ts (5.057 s)

Test Suites: 7 passed, 7 total
Tests:       97 passed, 97 total
Snapshots:   55 passed, 55 total
Time:        6.566 s
Ran all test suites.
$ cross-env NODE_ENV=production JEST_JUNIT_OUTPUT=../../test-results/mst-middlewares/prod.xml yarn jest
$ jest
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) The option `tsConfig` is deprecated and will be removed in ts-jest 27, use `tsconfig` instead
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
ts-jest[config] (WARN) Got `sourceMap: false` from tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json". This will disable source map support in ts-jest 27. Please set `sourceMap: true` in your tsconfig file "/Users/tylerwilliams/mst-middlewares/__tests__/tsconfig.json" if you want to keep source map support on
 PASS   test  __tests__/custom.test.ts
 PASS   test  __tests__/time-traveller.test.ts
 PASS   test  __tests__/action-logger.test.ts
 PASS   test  __tests__/simple-action-logger.test.ts
 PASS   test  __tests__/undo-manager.test.ts (5.182 s)
 PASS   test  __tests__/atomic.test.ts (5.864 s)
 PASS   test  __tests__/redux.test.ts (6.382 s)

Test Suites: 7 passed, 7 total
Tests:       95 passed, 95 total
Snapshots:   55 passed, 55 total
Time:        8.222 s
```
</details>

Reply:
The output does not change if I leave `call.env = undefined` in the `undo-manager` src as well.

Can you also let me know what commands you ran to get the output in the screenshot you sent me? Thanks!

Reply:
> The output does not change if I leave `call.env = undefined` in the `undo-manager` src as well.
> 
> Can you also let me know what commands you ran to get the output in the screenshot you sent me? Thanks!

after i checkout code. 
1 yarn install
2 yarn test（make sure things OK to go with）
3 write unit test (reported error )
4 do the real change.
…………
May be you forgot revert the original `call.env = undefiend`.
![image](https://github.com/coolsoftwaretyler/mst-middlewares/assets/9973072/716f82e0-704a-4f82-9f05-bcc5a45f1d9b)

Reply:
Ah yes! Very good catch, that's my mistake. I wasn't thinking in terms of a true revert. I can simulate the test failure now with that change, and everything is looking good.

I'm going to merge this PR, and then release a new version to npm. I will release it as a major verison upgrade in case anyone depends on the prior behavior. I think this is technically non-breaking, but since the library doesn't have a ton of tests, I'm going to err on the side of major version upgrades to keep people stable if they're automatically upgrading minor/patch versions out in the wild.

Thanks for the excellent work! I plan to get this done sometime today or later this week.